text/3634-scoped-impl-trait-for-type.md

```diff
@@ -0,0 +1,3436 @@
+
+As with `MonomorphicSubtrait`, the monomorphisation of `impl BoundedSubtrait for Type …` that is used in ❶ is shadowed in ❷.
+
+However, due to the `where Type: Trait` bound *on the implementation*, that implementation is polymorphic over `Trait for Type` implementations. This means a second monomorphisation is available in ❷ and its nested scope ❸.
+
+### `FnBoundedMonomorphic`
+
+`FnBoundedMonomorphic`'s implementations are monomorphic from the get-go just like `Monomorphic`'s.
+
+Due to the narrower bounds on functions, their availability can vary between receivers but always matches that of the global *implementation environment*:
+
+#### `::where_trait`
+
+Available everywhere since `Type: Trait` is in scope for both implementations of `FnBoundedMonomorphic`.
+
+In ❶, this resolves to the global implementation.
+
+In ❷, this *still* calls the global `<Type as Trait in ::>::function()` implementation since the global `FnBoundedMonomorphic` implementation is *not* polymorphic over `Type: Trait`.
+
+In ❸, `FnBoundedMonomorphic` is monomorphically reimplemented for `Type`, which means it "picks up" the scoped `Type: Trait` implementation that's in scope there from ❷.
+
+#### `::where_monomorphic_subtrait`
+
+In ❶, this resolves to the global implementation.
+
+In ❷, this *still* calls the global `<Type as MonomorphicSubtrait in ::>::monomorphic_subtrait()` implementation since the global `FnBoundedMonomorphic` implementation is *not* polymorphic over `Type: Trait`.
+
+Note that `FnBoundedMonomorphic` *cannot* be reimplemented in ❷ since the bound `Type: MonomorphicSubtrait` on its associated function isn't available in that scope, which would cause a difference in the availability of associated functions (which would cause a mismatch when casting to `dyn FnBoundedMonomorphic`).
+
+> It may be better to allow `use impl FnBoundedMonomorphic for Type {}` without `where_monomorphic_subtrait` in ❷ and disallow incompatible unsizing instead. I'm not sure about the best approach here.
+
+In ❸, `FnBoundedMonomorphic` is monomorphically reimplemented for `Type`, which means it "picks up" the scoped `Type: Trait` implementation that's in scope there from ❷.
+
+### `NestedMonomorphic` / `::nested_monomorphic`
+
+The global implementation of `NestedMonomorphic` in ❸ the binds to the scoped implementation of `Trait` on `Type` from ❷ internally. This allows outside code to call into that function indirectly without exposing the scoped implementation itself.
+
+### `BoundedOnOther` / `::bounded_on_other`
+
+As this discrete implementation's bound isn't over the `Self` type (and does not involved generics), it continues to act only as assertion and remains monomorphic.
+
+## Binding and generics
+
+`where`-clauses without generics or `Self` type, like `where (): Debug`, **do not** affect binding of implementations within an `impl` or `fn`, as the non-type-parameter-type `()` is unable to receive an *implementation environment* from the discretisation site.
+
```

Comment:
Yes, exactly that. (More precisely, such a scoped implementation *at the caller* does not affect the behaviour of the function, but *at its definition* it does, and the trivial bound is satisfied according to the function definition's implementation environment too.)

I think that, in this case, it's overall better to keep monomorphisation in check and not turn discrete-as-written implementations into monomorphised ones. Rustdoc *does* display trivial bounds though 🤔
(Another practical issue is inability to write implementation environments inline for anything that's not a generic type argument… but that could just be a case of me personally coming up blank if I think about the syntax for it.)

Changing this doesn't sit well with me because it seems unintuitive (it would make `where`-clauses a form of first-class implicit-only parameter) and there may be macros that emit spurious trivial bounds. It would also break that trivial bounds can be used as assertions, at least if allowed without global implementation… though the `static_assertions` crate notably does not currently do this.

In general, I'm very hesitant to add anything that reinterprets currently valid Rust code, even if I'm pretty sure this wouldn't be a compilation-breaking change. Maybe it would be a better idea to hide trivial bounds from the docs instead, going for consistency in the other direction?

Reply:
Makes sense to me, but you might consider explicitly calling out that trivial bounds are not affected by scoped implementations.